tiledb/sm/filesystem/s3.cc

```diff
@@ -222,6 +222,64 @@ class S3Exception : public StatusException {
   }
 };
 
+/**
+ * Helper class which overrides Aws::S3::S3Client to set a header from
+ * vfs.s3.curl_header
+ *
+ * @note vfs.s3.curl_header expects format "header_key header_value"
+ */
+class TileDBS3Client : public Aws::S3::S3Client {
+ public:
+  TileDBS3Client(
+      const S3Parameters& s3_params,
+      const Aws::S3::S3ClientConfiguration& client_config,
+      Aws::Client::AWSAuthV4Signer::PayloadSigningPolicy sign_payloads,
+      bool use_virtual_addressing)
+      : Aws::S3::S3Client(client_config, sign_payloads, use_virtual_addressing)
+      , params_(s3_params) {
+  }
+
+  TileDBS3Client(
+      const S3Parameters& s3_params,
+      const std::shared_ptr<Aws::Auth::AWSCredentialsProvider>& creds,
+      const Aws::S3::S3ClientConfiguration& client_config,
+      Aws::Client::AWSAuthV4Signer::PayloadSigningPolicy sign_payloads,
+      bool use_virtual_addressing)
+      : Aws::S3::S3Client(
+            creds, client_config, sign_payloads, use_virtual_addressing)
+      , params_(s3_params) {
+  }
+
+  virtual void BuildHttpRequest(
+      const Aws::AmazonWebServiceRequest& request,
+      const std::shared_ptr<Aws::Http::HttpRequest>& httpRequest)
+      const override {
+    S3Client::BuildHttpRequest(request, httpRequest);
+
+    // Set header from S3Parameters "header_key header_value"
+    auto header = params_.curl_header_;
+    if (header != "") {
+      if (header.find(" ") == std::string::npos) {
+        throw std::runtime_error(
+            "Failed to set curl header; vfs.s3.curl_header expects "
+            "a key-value pair in format \"header_key header_value\"");
+      }
+
+      httpRequest->SetHeaderValue(
+          header.substr(0, header.find(" ")),
+          header.substr(header.find(" ") + 1));
+    }
```

Comment:
Two things:

* The format `name value` does not support setting multiple custom headers. One format that would allow this would be `name1=value1;name2=value2`.
* Parsing the config value should be done at initialization time to save time and report the errors earlier, and the custom header names and values should be stored in an `std::vector<std::pair<std::string,std::string>`.

Reply:
The point about parsing at initialization time is completely right. Since parsing of all sorts is a whole category of security holes, the parser should have separate implementation and have unit tests. It's never "too simple to get wrong".

As for format, separators should be character sequences that are not found in the headers. Otherwise the specification won't be clean and require quoting to be so. This is preventative avoidance of a future support problem.

Colon is a good separator between field name and value, since that's the separator between field name and value. On the other hand, field values can contain most any single character, including colons and newlines, but they cannot contain a sequence of two newlines `\n\n`. So that makes a good delimiter for multiple headers.

The parser should include a citation for the HTTP header grammar: https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2